components/zcash_encoding/src/lib.rs

```diff
@@ -136,6 +148,46 @@ impl Vector {
         CompactSize::write(&mut writer, vec.len())?;
         vec.iter().try_for_each(|e| func(&mut writer, e))
     }
+
+    /// Writes at most `items_to_write` values from the provided iterator to the stream
+    /// in [`CompactSize`]-prefixed format.
+    ///
+    /// If not enough items are available, this will return an error; all available
+    /// elements will already have been written to the stream but the [`CompactSize`]
+    /// prefix that was written will be incorrect, and so the data written will not be
+    /// able to be correctly decoded as a vector.
```

Comment:
I do not like this as an API on `Vector`. My concern is not as much with not enough items being available (returning an error in this case is functionally equivalent to if the writing of an individual item failed; the IO process abruptly halts), but with too many items being available (and `items_to_write` being incorrectly set).

I would prefer a bound `I: Iterator<Item = E> + ExactSizeIterator`; then we remove the `items_to_write` parameter (instead using `items.len()`). Then if the caller explicitly wants to truncate their source of items, they can use `.take(items_to_write)` on the iterator passed in.